File: kotlin-semver/karakum/annotations/annotateJsPlainObject.js
import ts from "typescript"

const jsoInterfaces = [
    "RangeOptions",
    "Options",
    "CoerceOptions"
]

export default (node) => {
    if (
        ts.isInterfaceDeclaration(node)
        && jsoInterfaces.some(name => node.name.text === name)
    ) {
        return `@kotlinx.js.JsPlainObject`
    }

    return null
}

File: kotlin-semver/karakum/src/jsMain/kotlin/semver/karakum/annotations/annotateJsPlainObject.kt
package semver.karakum.annotations

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.AnnotationContext
import typescript.Node
import typescript.isInterfaceDeclaration

private val jsoInterfaces = setOf(
    "RangeOptions",
    "Options",
    "CoerceOptions"
)

fun annotateJsPlainObject(node: Node, context: AnnotationContext) = nullable {
    ensure(isInterfaceDeclaration(node))
    ensure(node.name.text in jsoInterfaces)

    "@kotlinx.js.JsPlainObject"
}
